overlord/devicestate/devicestate.go

```diff
@@ -620,6 +665,102 @@ func sideInfoAndPathFromID(sis []*snap.SideInfo, paths []string, id string) *pat
 	return nil
 }
 
+func modelValidationSetsToAsserts(m *asserts.Model, db asserts.RODatabase) ([]*asserts.ValidationSet, error) {
+	vsAsserts := make([]*asserts.ValidationSet, 0, len(m.ValidationSets()))
+	for _, vs := range m.ValidationSets() {
+		atSeq := vs.AtSequence()
+
+		a, err := resolveValidationSetAssertion(atSeq, db)
+		if err != nil {
+			return nil, fmt.Errorf("internal error: cannot resolve validation-set: %v", err)
+		}
+
+		vsAsserts = append(vsAsserts, a.(*asserts.ValidationSet))
+	}
+	return vsAsserts, nil
+}
+
+func resolveValidationSetAssertion(seq *asserts.AtSequence, db asserts.RODatabase) (asserts.Assertion, error) {
+	if seq.Sequence <= 0 {
+		hdrs, err := asserts.HeadersFromSequenceKey(seq.Type, seq.SequenceKey)
+		if err != nil {
+			return nil, err
+		}
+		return db.FindSequence(seq.Type, hdrs, -1, seq.Type.MaxSupportedFormat())
+	}
+	return seq.Resolve(db.Find)
+}
+
+func validationSetsFromModel(model *asserts.Model, st *state.State, store snapstate.StoreService, offline bool) (*snapasserts.ValidationSets, error) {
+	save := func(a asserts.Assertion) error {
+		return assertstate.Add(st, a)
+	}
+
+	errOfflineRemodel := errors.New("cannot fetch assertions during offline remodel")
+
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		if offline {
+			return nil, errOfflineRemodel
```

Comment:
I'm not sure this is entirely correct, I think this should return a db.Find instead from the database with all the pre-provided assertions?

Reply:
Shouldn't retrieveSeq and retrieve not be called in the case that they are already present in the db?

Reply:
Oh, no it doesn't work how I thought it did.